test/integration/controllermanager/credentialsbinding/referencecleaner/referencecleaner_test.go

```diff
@@ -0,0 +1,327 @@
+			Expect(testClient.Create(ctx, credentialsBinding1)).To(Succeed())
+			log.Info("Created CredentialsBinding for test", "credentialsBinding", client.ObjectKeyFromObject(credentialsBinding1))
+
+			By("Wait until manager has observed the CredentialsBinding")
+			Eventually(func() error {
+				return mgrClient.Get(ctx, client.ObjectKeyFromObject(credentialsBinding1), &securityv1alpha1.CredentialsBinding{})
+			}).Should(Succeed())
+
+			By("Create Secret")
+			Expect(testClient.Create(ctx, secret)).To(Succeed())
+			log.Info("Created Secret for test", "secret", client.ObjectKeyFromObject(secret))
+
+			DeferCleanup(func() {
+				By("Delete CredentialsBinding")
+				Expect(testClient.Delete(ctx, credentialsBinding1)).To(Or(Succeed(), BeNotFoundError()))
+				Eventually(func() error {
+					return testClient.Get(ctx, client.ObjectKeyFromObject(credentialsBinding1), credentialsBinding1)
+				}).Should(BeNotFoundError())
+
+				By("Delete Secret")
+				Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Or(Succeed(), BeNotFoundError()))
+				Expect(controllerutils.RemoveFinalizers(ctx, testClient, secret, "gardener.cloud/gardener")).To(Or(Succeed(), BeNotFoundError()))
+				Expect(testClient.Delete(ctx, secret)).To(Or(Succeed(), BeNotFoundError()))
+				Eventually(func() error {
+					return testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+				}).Should(BeNotFoundError())
+			})
+		})
+
+		Context("no CredentialsBinding referencing the Secret", func() {
+			It("should release the Secret", func() {
+				By("Ensure that Secret is not being deleted")
+				Consistently(func(g Gomega) {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+				}).Should(Succeed())
+
+				By("Delete the CredentialsBinding")
+				Expect(testClient.Delete(ctx, credentialsBinding1)).To(Succeed())
+
+				By("Ensure finalizer and labels are removed from Secret")
+				Eventually(func(g Gomega) {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+					g.Expect(secret.Finalizers).To(BeEmpty())
+					g.Expect(secret.Labels).ToNot(And(
+						HaveKey("provider.shoot.gardener.cloud/"+providerType),
+						HaveKey("reference.gardener.cloud/credentialsbinding"),
+					))
+				}).Should(Succeed())
+			})
+
+			It("should release the Secret but keep finalizer and provider type label because of SecretBinding ref label", func() {
+				secret.Labels["reference.gardener.cloud/secretbinding"] = "true"
+				Expect(testClient.Update(ctx, secret)).To(Succeed())
+
+				By("Ensure that secret is not being deleted")
+				Consistently(func(g Gomega) {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+				}).Should(Succeed())
+
+				By("Delete the CredentialsBinding")
+				Expect(testClient.Delete(ctx, credentialsBinding1)).To(Succeed())
+
+				By("Ensure finalizer and labels are removed from Secret")
```

Comment:
The test actually ensures the finalizer is not removed
```suggestion
				By("Ensure labels are removed from Secret")
```